### ISSUE TITLE
Cross entropy for packing

### ISSUE THREAD

**Issue by fzyzcjy** [linkedin/Liger-Kernel#329]
### 🚀 The feature, motivation and pitch

Hi thanks for the lib! The https://huggingface.co/blog/packing-with-FA2 new feature looks great and can further speed up training using this repo. However, it has some problem: if I understand correctly, the cross entropy loss will be different.

For example, suppose we have two sequences `[a,b,c,d,e,f,g]` and `[x,y]`. Then, the original way will give: 

```
loss = 1/2( 1/7(loss(a) + loss(b) + ... + loss(g)) + 1/2(loss(x) + loss(y)) )
```

while the new one - `[a,b,c,d,e,f,g,x,y]` will be

```
loss = 1/9(loss(a) + ... + loss(y))
```

As can be seen, for example, the "z" is originally weighted 1/4, but now weighted 1/9. 

It would be great if the cross entropy of liger kernel could handle this.

**Comment by pramodith** [comment]
From an initial exception, I think that the `cross_entropy` kernel would need to support an `is_packed` flag. We may not need to pass in the position ids since in Causal Language Modeling, we can figure out the boundaries just based on the `ignore_token_index` so all the location in the target/label tensor with the `ignore_index_token` will be the boundaries of a sequence.

#take @ByronHsu I think this'll be a good first issue for me to try and contribute to.

**Comment by fzyzcjy** [comment]
@pramodith Thanks!

> we can figure out the boundaries just based on the ignore_token_index so all the locations in the target/label tensor with the ignore_index_token will correspond to the boundaries of a sequence.

I am not very sure about it. For example, when SFT, we should train on the completions (i.e. the assistant answers) and make the question (i.e. the prompt) be masked and never have a gradient. Then we will see a lot of pad tokens for the prompt, while it is not a boundary.

**Comment by pramodith** [comment]
@fzyzcjy good point!
> I am not very sure about it. For example, when SFT, we should train on the completions (i.e. the assistant answers) and make the question (i.e. the prompt) be masked and never have a gradient. Then we will see a lot of pad tokens for the prompt, while it is not a boundary.

So let's say we have SFT training in (user, assistant) format as you said all the question tokens will be marked as pad/ignore tokens. So we might have a turn in the format [-100, -100, -100, 45, 67, 89, 88, -100], the thing is that as long as the ignore/pad tokens are contiguous we can treat them as a single boundary. 

Things can get admittedly harder if we are in a multi-turn conversation and we'd want to consider all the turns in a convo as a single conversation versus treating a single turn as a single sequence (post-packing). For an initial implementation I'd suggest going for the latter since its easier.

**Comment by pramodith** [comment]
I went through some of the code in Transformers to understand if they correctly implement the cross_entropy loss for packed sequences and I don't think that they're doing it either. 😞 

For example, this is the default loss function that training a CausalLM reverts to if the model's implementation doesn't explicitly mention a loss function (LLama for e.g.) https://github.com/huggingface/transformers/blob/913330ca9f80b0a308d7490a02274b01b51e6051/src/transformers/loss/loss_utils.py#L32

Over there they lose all information about a sequence when computing the loss because of the reshape operations before calling the `cross_entropy` function, liger kernels follow this style of reshaping to a 2d tensor before computing the loss too, as seen here: https://github.com/linkedin/Liger-Kernel/blob/1c0c75c3455e788d575966bfc5edec3ef166835e/src/liger_kernel/transformers/model/llama.py#L131

When, I dive into the Trainer API offered by Transformers, the default loss function the revert to for CausalLM is the `LabelSmoother` https://github.com/huggingface/transformers/blob/8a734ea2c340beee23e665601919814918bf4c43/src/transformers/trainer.py#L3690

Which is essentially a custom but more or less same implementation of the cross entropy function, but even they don't seem to take packing into consideration and keep track of the actual length of the packed sequences before aggregating the loss.

https://github.com/huggingface/transformers/blob/913330ca9f80b0a308d7490a02274b01b51e6051/src/transformers/trainer_pt_utils.py#L559

So not sure, if we should be supporting this in liger until HF does.

**Comment by fzyzcjy** [comment]
> Things can get admittedly harder if we are in a multi-turn conversation 

I guess so. It seems SFTTrainer does allow such multi-turn conversation. 

> For an initial implementation I'd suggest going for the latter since its easier. 

Sure, it looks like a pretty good initial implementation.

> I went through some of the code in Transformers to understand if they correctly implement the cross_entropy loss for packed sequences and I don't think that they're doing it either. 😞

Great catch!

**Comment by fzyzcjy** [comment]
@pramodith I created an issue at hf transformers to ask this: https://github.com/huggingface/transformers/issues/34510
